### ISSUE TITLE
power_assert.rb:8: warning: tracing (e.g. set_trace_func) will not capture all events without --debug flag

### ISSUE THREAD

**Issue by eregon** [jruby/jruby#7661]
`jruby 9.4.1.0 (3.1.0) 2023-02-07 237d5fa5f4 OpenJDK 64-Bit Server VM 11.0.18+10 on 11.0.18+10 +jit [x86_64-linux]`
See https://github.com/ruby/timeout/actions/runs/4187122427/jobs/7256547037
Due to this, JRuby fails *all* `assert_separately` tests.

**Comment by ahorek** [comment]
yes, multiple stdlib ruby repositories are affected

https://github.com/ruby/cgi/blob/fc56707d244d9e80472bb77e94a454f24e3e2390/test/lib/core_assertions.rb#L279 is a workaround, but probably not an ideal solution.

**Comment by eregon** [comment]
I guess the proper fix for this is in JRuby or in the `power_assert`, right?
So that the feature check doesn't emit warnings as a side effect.

**Comment by headius** [comment]
Depending on which events are requested, we may be warning too frequently. Because many JRuby execution modes do not use invokedynamic, it's difficult for us to support tracing without also impacting performance. Some events are trivial to support, like exception raising or opening a class, but others are invasive like line events. We introduced the --debug flag to enable these events only when needed.

I this case, it seems like power_assert wants to hook `return` and `c_return` events. In order for these to be always on, we'd need to have code surrounding every dynamic call that calls back into the hook mechanism, whether a hook is present or not. That's easy to do with invokedynamic, since it would eliminate the hook call if no hook is enabled, but the other modes would need to make this check every time, impacting non-hooked performance.

There's also simpler options:

1. Simply pass --debug to JRuby for these test runs. That will enable the events in question across all code, eliminate the warning, and also allow the full range of hook-based tests to function properly. This is what we do on a limited basis in the JRuby CI runs.
2. Modify power_assert to check the engine or query for capabilities before enabling these hooks. We would lose some reporting, perhaps, but I'm not sure how big of an impact it would be. I have not fully explored what these hooks are being used for...there may be another way to achieve the same effect.

**Comment by eregon** [comment]
I think we should really avoid needing to set `JRUBY_OPTS` or so in every workflow using JRuby.
So a solution in `power_assert` seems best.
I'm also unsure what `power_assert` uses TracePoint for, and it seems `test-unit` requires `power_assert` eagerly even though it might not be used. So maybe an obvious fix is to feature-check lazily, on the first use of `assert { ... }` (seems very rarely used).
Also feature checking by creating a global TracePoint is extremely expensive (rather it should target a specific method to avoid that overhead), so something to avoid anyway.

**Comment by eregon** [comment]
> That's easy to do with invokedynamic, since it would eliminate the hook call if no hook is enabled, but the other modes would need to make this check every time, impacting non-hooked performance.

Another solution would be to always use invokedynamic for TracePoint, IIRC you have said before some invokedynamic things are used even without the flag (constants maybe?).

BTW, TruffleRuby fails early the feature check here and doesn't have this issue because `:c_return` is not implemented. I suppose JRuby doesn't implement `:c_return` either so another solution would be to just raise `unknown event: c_return (ArgumentError)`.

It is likely better for performance on JRuby/non-CRuby if there is no TracePoint hook actually used, i.e., if https://github.com/ruby/power_assert/blob/7919ed821cb494bd51c52c6596e7675f06a35239/lib/power_assert.rb#L17 happens. So maybe a `RUBY_ENGINE == "ruby"` could also make sense since `power_assert` seems optional anyway (at least when using `test-unit`).

**Comment by headius** [comment]
> you have said before some invokedynamic things are used even without the flag (constants maybe?).

Yes, constants, thread event polling, a few other things. But calls are not, and in order to properly handle call/return events they'd need to be hooked at call time.

I have not looked into this in a while, but I am still pushing toward invokedynamic being always on. It's just hard to do without impacting startup time.

> I suppose JRuby doesn't implement :c_return

JRuby does impleent `:c_return`:

```
$ jruby --debug -e 'set_trace_func ->(*args){ p args}; "foo".upcase'
["c-call", "-e", 1, :upcase, #<Binding:0x7161457>, nil]
["c-return", "-e", 1, :upcase, #<Binding:0x629984eb>, nil]
```

But of course as in TR many methods MRI implements in C are implemented in Ruby for us, so it's not always going to match MRI.

**Comment by headius** [comment]
I looked into `call` and `return` events a bit to refresh my memory on how I implemented them.

The `c_call` and `c_return' events are implemented inside the function "pointers" used to bind Java methods, and normally we pregenerate those pointers without that hook to avoid the overhead. Enabling debug mode skips the pregenerated pointers and generates new ones with the hook calls in place. The generated pointers are used by all compilation modes and could possibly use indy.

On the other hand, "call", and "return" are implemented in IR, as instructions within Ruby-based methods on entry and exit. We only emit these instructions when debug mode is on. In non-indy JIT mode they are emitted as normal Java calls to the hook callback. In indy mode, they are emitted as a safe point check.

It would be interesting to experiment with using indy for c_call and c_return, since they'd be zero-cost when no hooks are in use and the bytecode size for the function pointers is already quite small. And if we can get to always-on invokedynamic in the JIT, the `call` and `return` events would be similarly lightweight.

In the meantime, however, it would be best to just eliminate or reduce the use of these traces during testing. I'm still not sure why they are needed.

**Comment by headius** [comment]
After studying power_assert a bit more, I understand how tracing is used: to capture the return value from each call that happens in the assert block. From the readme, this is used to produce an error report showing each method call's return value:

```
    Failure:
       assert { 3.times.to_a.include?(3) }
                  |     |    |
                  |     |    false
                  |     [0, 1, 2]
                  #<Enumerator: 3:times>
```

It uses the trace events plus a reparse of the source in question (using Ripper) to find the offsets of these columns. So both the trace events and Ripper must be supported in order for it to produce this output correctly.

I have a WIP patch that enables always-on `call`, `return`, `b_call`, and `b_return` using indy and have confirmed that the eventual native code is unchanged if there are no hooks installed. Getting `c_call` and `c_return` to do the same will be trickier, since they do not have all Ruby execution data available (no way to access file or line number currently).

**Comment by headius** [comment]
Connecting the dots... the PR in question was #7673, but as mentioned in https://github.com/jruby/jruby/pull/7673#issuecomment-1468778798 we did not ultimately enable the call tracing by default.
